### PR TITLE
initial contribution

### PR DESCRIPTION
**Pull request by arnisoph**

### CODE REVIEW

File: configmap.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	corev1 "github.com/ericchiang/k8s/apis/core/v1"
+)
+
+func writeFile(file *string, contents *string) error {
+	os.MkdirAll(filepath.Dir(*file), 0755)
```
Comment (hikhvar) [inline configmap.go]: Error not checked

File: configmap.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	return ioutil.WriteFile(*file, []byte(*contents), 0644)
+}
+
+func writeOSFile(file *os.File, contents *string) (int, error) {
+	os.MkdirAll(filepath.Dir(file.Name()), 0755)
```
Comment (hikhvar) [inline configmap.go]: Error not checked. Don't use a pointer to the string content. Rather use []byte.

File: configmap.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	return file.Write([]byte(*contents))
+}
+
+func verifyCM(configMap *corev1.ConfigMap, verifySteps *[]verifyStep) (map[string]string, string, error) {
```
Comment (hikhvar) [inline configmap.go]: Don't use *[], simply use [].

File: configmap.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+				tempFile, err := ioutil.TempFile("", fmt.Sprintf("trovilo-%s-", file))
+				if err != nil {
+					return verifiedFiles, "", err
+				}
+				_, err = writeOSFile(tempFile, &fileContents)
+				if err != nil {
+					return verifiedFiles, "", err
+				}
+
+				for cmdPos := range (*verifySteps)[step].Cmd {
```
Comment (hikhvar) [inline configmap.go]: for _, arg := range varifySteps[step].Cmd

File: configmap.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+					} else {
+						args = append(args, arg)
+					}
+				}
+				cmd := exec.Command(args[0], args[1:]...)
+				out, err := cmd.CombinedOutput()
+				niceOutput := strings.TrimSpace(string(out))
+
+				// Remove the file, regardless of the verification result
+				os.Remove(tempFile.Name())
```
Comment (hikhvar) [inline configmap.go]: use defer os.Remove(tempFile.Name()) After creation of the tempFile

File: configmap.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+				verifiedFiles[file] = niceOutput
+			}
+		}
+
+	}
+
+	return verifiedFiles, "", nil
+}
+
+func compareCMLabels(expected *map[string]string, actual *map[string]string) bool {
```
Comment (hikhvar) [inline configmap.go]: Pointer auf Maps

File: configmap.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	for key, expectedValue := range *expected {
+		actualValue, found := (*actual)[key]
+		if !found || expectedValue != actualValue {
+			return false
+		}
+	}
+	return true
+}
+
+func registerCM(configMap *corev1.ConfigMap, targetDir *string) ([]string, error) {
```
Comment (hikhvar) [inline configmap.go]: Pointer

File: main.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+					log.WithFields(logrus.Fields{
+						"job":       config.Jobs[jobPos].Name,
+						"configmap": *cm.Metadata.Name,
+						"namespace": *cm.Metadata.Namespace,
+					}).Info("ConfigMap has been deleted from namepace, thus removing in target directory too")
+
+					removedFiles, err := removeCMfromTargetDir(cm, &config.Jobs[jobPos].TargetDir)
+
+					if err == nil {
+						log.WithFields(logrus.Fields{
```
Comment (hikhvar) [inline main.go]: Duplicate code.

logEntry := log.WithFields(logrus.Fields{
							"job":          config.Jobs[jobPos].Name,
							"configmap":    *cm.Metadata.Name,
							"namespace":    *cm.Metadata.Namespace,
							"removedFiles": removedFiles,
						})

logEntry.WithError(err).Fatal...
logEntry.Info()

File: main.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+					"configmap":   *cm.Metadata.Name,
+					"namespace":   *cm.Metadata.Namespace,
+					"verifySteps": config.Jobs[jobPos].Verify,
+				}).Debug("Verifying ConfigMap against validity")
+
+				// Verify validity of ConfigMap files
+				verifiedFiles, latestOutput, err := verifyCM(cm, &config.Jobs[jobPos].Verify)
+				if err != nil {
+					log.WithFields(logrus.Fields{
+						"job":           config.Jobs[jobPos].Name,
```
Comment (hikhvar) [inline main.go]: Duplicate Logs

File: util.go [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+func setupLogging() {
+	if *logJSON {
+		log.Formatter = new(logrus.JSONFormatter)
+	}
+
+	log.Out = os.Stdout
+
+	if *logLevel == "debug" {
```
Comment (hikhvar) [inline util.go]: Switch case

File: configmap.go (hunk 1 above)
Comment (arnisoph) [reply configmap.go]: done 👍

File: configmap.go (hunk 2 above)
Comment (arnisoph) [reply configmap.go]: done 👍

File: configmap.go (hunk 3 above)
Comment (arnisoph) [reply configmap.go]: done 👍

File: configmap.go (hunk 4 above)
Comment (arnisoph) [reply configmap.go]: done

File: configmap.go (hunk 5 above)
Comment (arnisoph) [reply configmap.go]: done 👍

File: configmap.go (hunk 6 above)
Comment (arnisoph) [reply configmap.go]: done 👍

File: configmap.go (hunk 7 above)
Comment (arnisoph) [reply configmap.go]: done 👍

File: main.go (hunk 8 above)
Comment (arnisoph) [reply main.go]: done 👍

File: main.go (hunk 9 above)
Comment (arnisoph) [reply main.go]: done 👍

File: util.go (hunk 10 above)
Comment (arnisoph) [reply util.go]: done 👍

**Comment by arnisoph** [comment]
ready4review again. I will implement the post actions (http/cmd) via a dedicated PR. Thx so far!

File: util/util.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
+package util
```
Comment (hikhvar) [inline util/util.go]: Personally, I would call this package logging instead of util. It handles the logging setup.

**Review by hikhvar — COMMENTED** [review]
Execpt minor naming issues. LGTM

File: util/util.go (hunk 11 above)
Comment (arnisoph) [reply util/util.go]: done!

**Comment by arnisoph** [comment]
just FYI, this branch doesn't contain CI files yet

**Review by hikhvar — APPROVED** [review]

**Pull request merged**